src/libertem/executor/pipelined.py

```diff
@@ -0,0 +1,901 @@
+                        del id_to_task[result_task_id]
+                        assert len(id_to_task) == in_flight
+                except WorkerQueueEmpty:
+                    continue
+
+            # FIXME: code duplication
+            # at the end, block to get the remaining results:
+            while in_flight > 0:
+                try:
+                    with self._pool.response_queue.get() as (result, _):
+                        in_flight -= 1
+                        if result["type"] == "ERROR":
+                            _raise_from_msg(result, "failed to run tasks")
+                        result_task_id = result["task_id"]
+                        yield (result["result"], id_to_task[result_task_id], result_task_id)
+                        del id_to_task[result_task_id]
+                        assert len(id_to_task) == in_flight
+                except WorkerQueueEmpty:
+                    continue
+
+            task_comm_handler.done()
+        except Exception:
+            # In case of an exception, we need to drain the response queue,
+            # so the next `run_tasks` call isn't polluted by old responses.
+            # -> just like in the happy case, we need to wait for responses
+            # for all in-flight tasks. In case the error happened between incrementing
+            # `in_flight` and actually sending the task to the queue, we should
+            # have a timeout here to not wait infinitely long.
+            while in_flight > 0:
+                try:
+                    with self._pool.response_queue.get(
+                        timeout=self._cleanup_timeout
+                    ) as (result, _):
+                        in_flight -= 1
+                        # we only raise the first exception; log the others here:
+                        if result["type"] == "ERROR":
+                            logger.error(f"Error response from worker: {result['error']}")
+                except WorkerQueueEmpty:
+                    continue
+            # if from a worker, this is the first exception that got put into the queue
+            raise
+
+    def run_tasks(
+        self,
+        tasks: Iterable[TaskProtocol],
+        params_handle: Any,
+        cancel_id: Any,
+        task_comm_handler: "TaskCommHandler",
+    ) -> ResultT:
+        with tracer.start_as_current_span("PipelinedExecutor.run_tasks"):
+            yield from _order_results(self._run_tasks_inner(
+                tasks, params_handle, cancel_id, task_comm_handler,
+            ))
+
+    def get_available_workers(self) -> WorkerSet:
+        resources_by_kind = {
+            "CPU": {"compute": 1, "CPU": 1, "ndarray": 1},
+            "CUDA": {"compute": 1, "CUDA": 1},
+        }
+
+        def _resources_for_spec(worker_spec: WorkerSpec):
+            resources = resources_by_kind[worker_spec["device_kind"]]
+            if worker_spec["has_cupy"]:
+                resources["ndarray"] = 1
+            return resources
```

Comment:
This looks correct to me. Once the executor can map tasks to the eligible workers we can see if it passes the corresponding tests of the `DaskJobExecutor`.